Explore: Fix showing of full log context

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
This PR fixes the issue with cropped log context for the first and last ~10 lines. Moreover, it make "Show context" button more accessible for unwrapped logs by pushing it to second line instead of the having it at the end of the line.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/27262
Fixes https://github.com/grafana/grafana/issues/32870

**Special notes for your reviewer**:

This PR tackles 3 areas:
1.  Add `display: block` for the `Show/Hide context` button for unwrapped lines. This way, it is always visible and accessible for users. Otherwise, users would have to scroll to the end of the line to discover this button & feature.
2. For wrapped logs, we are fixing the issue with cropped log context by adding following styling below. The issue in this case was the fact, that if `overflow-x` would be always `scroll` (even for wrapped logs that are not scrolled), `overflow-y` would be always switched to `auto`, as combination of `scroll` and `visible` overflows is not possible. You can learn more here: https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-y#syntax 
```
      overflow-x: ${wrapLogMessage ? 'unset' : 'scroll'};
      overflow-y: visible;
```

3. For unwrapped logs, we need to set `overflow-x: scroll`. Therefore solution above won't work. This was tricky to solve and I have decided to use a solution, where we remove relative position on log line and use the relative positioning of log container, that doesn't have use `overflow: scroll` styling. And then just position log context by margins. This works nicely because non-wrapped logs have always 1 line of log and 1 line of Show/Hide context button, so the correct position can be calculated and achieved by using margins.


**Non-wrapped log lines:**
![image](https://user-images.githubusercontent.com/30407135/127837087-c8f15654-b46a-4e24-a95c-372714984415.png)


**Wrapped log lines:**
![image](https://user-images.githubusercontent.com/30407135/127836425-c53f77b5-a309-4e4d-a3d7-044e802cce95.png)

Another options that I considered and decided not to go with:
- Using Portal - with Portal, if you scroll, your context will be out of place. You could only display 1 context.
- Using javascript - We would have to do calculations for wrapped/unwrapped log lines. If we wanted to use this in LogsPanel, it could be more complicated than when we are using css.

ivanahuckova committed:
Fix log context

ivanahuckova committed:
Update to auto values insted of unset

ivanahuckova committed:
Update comment

ifrost commented:
Show/hide button looks a bit misplaced (shifted to the right)?

<img width="783" alt="Screen Shot 2021-08-02 at 12 37 10" src="https://user-images.githubusercontent.com/745532/127848831-65d7f97a-0c1b-4d12-87f3-8705aaa31c54.png">

BTW. any reason why show context is not a button in-lined with the entire line? Something like 2x, 3x for de-duping. With unwrapped lines it adds a lot of extra space that is filled in only on hover.

ivanahuckova commented:
>Show/hide button looks a bit misplaced (shifted to the right)?

True, 10px margin left should be removed.

> BTW. any reason why show context is not a button in-lined with the entire line? Something like 2x, 3x for de-duping. With unwrapped lines it adds a lot of extra space that is filled in only on hover.

Not sure if you've seen this issue https://github.com/grafana/grafana/issues/32870. Problem is that if you have it at the end, you can't see it with the long lines. If you've had it at the beginning, we would have extra space at the beginning which would look awkward. Another option would be to add show context only on hover, but when I tried it, it was annoying as the log lines were jumping around. But if you think that we shouldn't do this, I can remove this fix and just keep the context-visibility-related one.

ifrost commented:
> > Show/hide button looks a bit misplaced (shifted to the right)?
> 
> True, 10px margin left should be removed.
> 
> > BTW. any reason why show context is not a button in-lined with the entire line? Something like 2x, 3x for de-duping. With unwrapped lines it adds a lot of extra space that is filled in only on hover.
> 
> Not sure if you've seen this issue #32870. Problem is that if you have it at the end, you can't see it with the long lines. If you've had it at the beginning, we would have extra space at the beginning which would look awkward. Another option would be to add show context only on hover, but when I tried it, it was annoying as the log lines were jumping around. Another option would be adding on hover button, but that would be covering the text. But if you think that we shouldn't do this, I can remove this fix and just keep the context-visibility-related one. We can pass this to UX to help us solve it.

I think now it's way better now because it's finally usable :D But I'd run it by UX anyway because this way it also looks awkward with lots of extra empty space. I think we could merge it as it is and review it later. It's definitely an improvement to the usability.

ivanahuckova committed:
Fix redundant left margin for non-wrapped logs

ivanahuckova commented:
@ifrost I have pushed the fix for left margin.

ivanahuckova commented:
+cc @grafana/grafana-labs-ux-design

ifrost commented on packages/grafana-ui/src/components/Logs/LogRowContext.tsx:
```diff
@@ -13,25 +13,47 @@ import { LogMessageAnsi } from './LogMessageAnsi';
       box-shadow: 0 0 10px ${theme.colors.dropdownShadow};
       border: 1px solid ${theme.colors.bg2};
       border-radius: ${theme.border.radius.md};
+      width: ${wrapLogMessage ? 'calc(100% + 20px)' : 'auto'};
```
`commonStyles` are used only in `LogRowContextGroup` but without `wrapLogMessage` being passed (it uses `useStyle()`). Does it mean width/left could be just set here to `auto` or we should pass `wrapLogMessage` when getting these styles?

ivanahuckova replied:
Good catch. It seems that both of these can be removed (and their default value will be set to auto).

ivanahuckova committed:
Remove unused css styling

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/fix-context

ivanahuckova committed:
Adjust width and height

ivanahuckova commented:
@ifrost in https://github.com/grafana/grafana/pull/37442/commits/6e952179d937bb0162597eb746093a3857587519 I have also adjusted width so if the logs are long, it is not overflowing container and added min height to custom scrollbar, so the horizontal scrollbar is at the bottom of context container.
